packages/accounts-controller/src/AccountsController.ts

```diff
@@ -206,10 +222,33 @@ export class AccountsController extends BaseController<
   /**
    * Returns an array of all internal accounts.
    *
+   * @param chainIdOrNamespace - The chain ID or namespace.
    * @returns An array of InternalAccount objects.
    */
-  listAccounts(): InternalAccount[] {
-    return Object.values(this.state.internalAccounts.accounts);
+  listAccounts(
+    chainIdOrNamespace?: CaipChainId | CaipNamespace,
+  ): InternalAccount[] {
+    const accounts = Object.values(this.state.internalAccounts.accounts);
+    if (!chainIdOrNamespace) {
+      return accounts;
+    }
+
+    if (chainIdOrNamespace === ETHEREUM_CAIP_ID_WILDCARD) {
+      return accounts.filter((account) =>
+        account.type.startsWith(ETHEREUM_CAIP_NAMESPACE),
+      );
+    }
+
+    if (
+      !isCaipChainId(chainIdOrNamespace) &&
+      chainIdOrNamespace !== ETHEREUM_CAIP_ID_WILDCARD
+    ) {
+      throw new Error(`Invalid CAIP2 id ${String(chainIdOrNamespace)}`);
+    }
```

Comment:
We should probably throw the error as soon as possible to avoid executing redundant logic